File: src/main/java/seremis/geninfusion/soul/allele/AlleleDoubleArray.java
package seremis.geninfusion.soul.allele;

import net.minecraft.nbt.NBTTagCompound;
import seremis.geninfusion.api.soul.EnumAlleleType;
import seremis.geninfusion.soul.Allele;

public class AlleleDoubleArray extends Allele {

    public double[] value;

    public AlleleDoubleArray(boolean isDominant, double[] value) {
        super(isDominant, EnumAlleleType.DOUBLE_ARRAY);
        this.value = value;
    }

    public AlleleDoubleArray(Object... args) {
        super(args);
        value = new double[args.length - 1];
        for(int i = 1; i < args.length; i++) {
            value[i - 1] = (Float) args[i];
        }
    }

    public AlleleDoubleArray(NBTTagCompound compound) {
        super(compound);
        readFromNBT(compound);
    }

    @Override
    public void writeToNBT(NBTTagCompound compound) {
        super.writeToNBT(compound);
        compound.setInteger("length", value.length);
        for(int i = 0; i < value.length; i++) {
            compound.setDouble("value" + i, value[i]);
        }
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        super.readFromNBT(compound);
        value = new double[compound.getInteger("length")];
        for(int i = 0; i < value.length; i++) {
            value[i] = compound.getDouble("value" + i);
        }
    }

}

File: src/main/java/seremis/geninfusion/soul/allele/AlleleDoubleArray.scala
package seremis.geninfusion.soul.allele

import net.minecraft.nbt.NBTTagCompound
import seremis.geninfusion.api.soul.EnumAlleleType
import seremis.geninfusion.soul.Allele

class AlleleDoubleArray(isDominant: Boolean, var value: Array[Double]) extends Allele(isDominant, EnumAlleleType.DOUBLE_ARRAY) {

    def this(args: AnyRef*) {
        this(args(0).asInstanceOf[Boolean], args(1).asInstanceOf[Array[Double]].drop(1))
    }

    def this(compound: NBTTagCompound) {
        this(false, null)
        readFromNBT(compound)
    }

    override def writeToNBT(compound: NBTTagCompound) {
        super.writeToNBT(compound)
        compound.setInteger("length", value.length)
        for (i <- 0 until value.length) {
            compound.setDouble("value" + i, value(i))
        }
    }

    override def readFromNBT(compound: NBTTagCompound) {
        super.readFromNBT(compound)
        value = Array.ofDim[Double](compound.getInteger("length"))
        for (i <- 0 until value.length) {
            value(i) = compound.getDouble("value" + i)
        }
    }
}
